### PR TITLE
[Merged by Bors] - Move redeclaration errors to parser

### PR DESCRIPTION
**Pull request by raskad**
This Pull Request changes the following:

- Implement redeclaration errors in the parser
- Remove redeclaration errors from the compiler (this is a big step towards #1907)
- Fix some failing tests on the way

This requires a slight change in our public api. The Parser new requires a full `Context` instead of just the `Interner` for parsing new code. This is required, because if multiple scripts are parsed (e.g. every input in the REPL) global variables must be checked for redeclarations.

### CODE REVIEW

**Review by HalidOdat — APPROVED** [review]
This looks good from my side! :)

**Comment by raskad** [comment]
Rebased

File: boa_engine/src/syntax/ast/node/iteration/mod.rs [hunk 1]
```diff
@@ -26,6 +26,22 @@ pub enum IterableLoopInitializer {
     DeclarationPattern(DeclarationPattern),
 }
 
+impl IterableLoopInitializer {
+    pub(crate) fn bound_names(&self) -> Vec<Sym> {
```
Comment (Razican) [inline boa_engine/src/syntax/ast/node/iteration/mod.rs:30]: Could we add some documentation to this function?

File: boa_engine/src/syntax/ast/node/statement_list/mod.rs [hunk 2]
```diff
@@ -68,64 +68,86 @@ impl StatementList {
...
-                                );
-                            }
-                        }
-                        Declaration::Pattern(p) => {
-                            for ident in p.idents().iter().copied() {
-                                if !set.insert(ident) {
-                                    unreachable!(
-                                        "Redeclaration of {}",
-                                        interner.resolve_expect(ident)
+    pub(crate) fn lexically_declared_names(&self) -> Vec<(Sym, bool)> {
```
Comment (Razican) [inline boa_engine/src/syntax/ast/node/statement_list/mod.rs:71]: Could we add some documentation to this function?

File: boa_engine/src/syntax/ast/node/statement_list/mod.rs [hunk 3]
```diff
@@ -68,64 +68,86 @@ impl StatementList {
...
 
-    pub fn var_declared_names(&self) -> FxHashSet<Sym> {
-        let mut set = FxHashSet::default();
-        for stmt in self.items() {
-            if let Node::VarDeclList(decl_list) = stmt {
-                for decl in decl_list.as_ref() {
-                    match decl {
-                        Declaration::Identifier { ident, .. } => {
-                            set.insert(ident.sym());
+    pub(crate) fn lexically_declared_names_top_level(&self) -> Vec<Sym> {
```
Comment (Razican) [inline boa_engine/src/syntax/ast/node/statement_list/mod.rs:116]: Could we add some documentation to this function?

File: boa_engine/src/syntax/ast/node/statement_list/mod.rs [hunk 4]
```diff
@@ -68,64 +68,86 @@ impl StatementList {
...
+                },
+                _ => {}
             }
         }
-        set
+
+        names
+    }
+
+    pub(crate) fn var_declared_names_new(&self, vars: &mut FxHashSet<Sym>) {
```
Comment (Razican) [inline boa_engine/src/syntax/ast/node/statement_list/mod.rs:147]: This function could benefit from some documentation too.

**Review by Razican — APPROVED** [review]
I'm happy to merge it as-is, looks very good and this is a great improvement now! Check my comments to see if it makes sense to take action on them :)

**Pull request closed without merging**